### PR TITLE
Add pure fp16 training.

### PR DESCRIPTION
**Pull request by wzzju**
Add pure FP16 training with master weights (multi-precision) and rescale_grad. After training 120 passes on ResNet50 model, the top-1 accuracy and top-5 accuracy is described as below.
|FP32 or AMP|the num of epoch| top-1 accuracy| top-5 accuracy|
|:---|:---|:---|:---|
|FP32|120|75.674%|92.674%|
|Pure FP16|120|75.108%|92.303%|

* Training loss consistency verification during the whole training process.
image

* Average loss and test accuracy consistency verification after each epoch.
image
image
image
image
image
image

### CODE REVIEW

**Review by phlrain — APPROVED** [review]

File: PaddleCV/image_classification/build_model.py [hunk 1]
```diff
@@ -36,15 +35,21 @@ def _basic_model(data, model, args, is_train):
...
+        image_transpose.stop_gradient = image.stop_gradient
+        net_out = model.net(input=image_transpose,
                             class_dim=args.class_dim,
                             data_format=args.data_format)
     else:
         net_out = model.net(input=image, class_dim=args.class_dim)
-    softmax_out = fluid.layers.softmax(net_out, use_cudnn=False)
+    if args.use_pure_fp16:
+        net_out_fp32 = fluid.layers.cast(x=net_out, dtype="float32")
+        softmax_out = fluid.layers.softmax(net_out_fp32, use_cudnn=False)
```
Comment (Xreki) [inline PaddleCV/image_classification/build_model.py]: softmax为啥要设置use_cudnn=False？另外你这个代码也可以简化一下。

File: PaddleCV/image_classification/build_model.py [hunk 2]
```diff
@@ -96,15 +102,21 @@ def _mixup_model(data, model, args, is_train):
     lam = data[3]
 
     if args.model == "ResNet50":
-        image_in = fluid.layers.transpose(
-            image, [0, 2, 3, 1]) if args.data_format == 'NHWC' else image
-        image_in.stop_gradient = image.stop_gradient
-        net_out = model.net(input=image_in,
+        image_data = (fluid.layers.cast(image, 'float16')
```
Comment (Xreki) [inline PaddleCV/image_classification/build_model.py:102]: cast现在是能自动传递stop_gradient了吗？你这个image_data是没有设置stop_gradient的吧，这个cast不会自动插入反向了？

File: PaddleCV/image_classification/build_model.py [hunk 3]
```diff
@@ -114,11 +126,17 @@ def _mixup_model(data, model, args, is_train):
...
+    if args.use_pure_fp16:
+        target_loss_a = fluid.layers.reduce_sum(x=loss_a)
+        target_loss_b = fluid.layers.reduce_sum(x=loss_b)
+        cost = lam * target_loss_a + (1 - lam) * target_loss_b
+        target_cost = fluid.layers.reduce_sum(x=cost)
+    else:
+        target_loss_a = fluid.layers.mean(x=loss_a)
+        target_loss_b = fluid.layers.mean(x=loss_b)
+        cost = lam * target_loss_a + (1 - lam) * target_loss_b
+        target_cost = fluid.layers.mean(x=cost)
```
Comment (Xreki) [inline PaddleCV/image_classification/build_model.py:135]: 代码是不是可以再简化一下？

File: PaddleCV/image_classification/scripts/train/ResNet50_fp16.sh [hunk 4]
```diff
@@ -8,7 +8,9 @@ export FLAGS_cudnn_batchnorm_spatial_persistent=1
 DATA_DIR="Your image dataset path, e.g. /work/datasets/ILSVRC2012/"
 
 DATA_FORMAT="NHWC"
-USE_FP16=true #whether to use float16
+USE_AMP=true #whether to use amp
+USE_PURE_FP16=false
+MULTI_PRECISION=${USE_PURE_FP16}
```
Comment (Xreki) [inline PaddleCV/image_classification/scripts/train/ResNet50_fp16.sh:13]: multi_precision不可以直接通过use_pure_fp16来判断吗？另外既然有个use_amp，那use_pure_fp16是不是也可以像mxnet一样，直接用use_fp16？

参数太多，容易让人迷惑。

File: PaddleCV/image_classification/utils/optimizer.py [hunk 5]
```diff
@@ -159,9 +160,22 @@ def __init__(self, args):
...
         self.decay_rate = args.decay_rate
         self.total_images = args.total_images
+        self.multi_precision = args.multi_precision
+        self.rescale_grad = (1.0 / (args.batch_size / self.get_gpu_num())
+            if args.use_pure_fp16 else 1.0)
 
         self.step = int(math.ceil(float(self.total_images) / self.batch_size))
 
+    def get_gpu_num(self):
+        visible_device = os.getenv('CUDA_VISIBLE_DEVICES')
```
Comment (Xreki) [inline PaddleCV/image_classification/utils/optimizer.py]: paddle有个接口是`core.get_cuda_device_count()`，不过core接口不建议使用了。可以用`len(fluid.cuda_places())`来获取设备数。你用环境变量获取，分布式训练的时候，可能会有问题吧。

File: PaddleCV/image_classification/build_model.py (hunk 1 above)
Comment (wzzju) [reply PaddleCV/image_classification/build_model.py]: Done.

File: PaddleCV/image_classification/build_model.py (hunk 2 above)
Comment (wzzju) [reply PaddleCV/image_classification/build_model.py:102]: 我再确认一下，因为后面transpose操作之后设置了stop_gradient，感觉这里不需要了吧。
image

File: PaddleCV/image_classification/build_model.py (hunk 3 above)
Comment (wzzju) [reply PaddleCV/image_classification/build_model.py:135]: 请教这里如何简化呢？

File: PaddleCV/image_classification/scripts/train/ResNet50_fp16.sh (hunk 4 above)
Comment (wzzju) [reply PaddleCV/image_classification/scripts/train/ResNet50_fp16.sh:13]: 即使使用纯fp16训练，multi_precision也是可选的。我个人觉得，use_pure_fp16和use_amp更容易区分，而use_fp16和use_amp才容易使人迷惑。

File: PaddleCV/image_classification/utils/optimizer.py (hunk 5 above)
Comment (wzzju) [reply PaddleCV/image_classification/utils/optimizer.py]: Done.

File: PaddleCV/image_classification/build_model.py (hunk 2 above)
Comment (wzzju) [reply PaddleCV/image_classification/build_model.py:102]: cast OP现在会自己传递stop_gradient了。
image

File: PaddleCV/image_classification/scripts/train/ResNet50_fp16.sh (hunk 4 above)
Comment (phlrain) [reply PaddleCV/image_classification/scripts/train/ResNet50_fp16.sh:13]: 好像纯的fp16，不开multi precision 也不work的
